apps/spruce/src/utils/tasks/estimatedActivatedTasks.ts

```diff
@@ -0,0 +1,106 @@
+import { VariantTask } from "gql/generated/types";
+import { VariantTasksState } from "hooks/useConfigurePatch";
+import {
+  selectedStrings,
+  versionSelectedTasks,
+} from "hooks/useVersionTaskStatusSelect";
+
+type ValidInputTypes = Set<string> | versionSelectedTasks | VariantTasksState;
+
+export const getNumEstimatedActivatedTasks = (
+  inputType: ValidInputTypes,
+  map: { [key: string]: number },
```

Comment:
The discrepancies in the multiple implementations of counting the number of estimated generated tasks stems from the fact that the `ConfigurePatchCore`, `VersionRestartModal`, and `ScheduleTasksModal` all storing the state of selected tasks in slightly different data structures. For example:

```
// VersionRestartModal. True indicates the task has been selected
{
  "example_version_id": {
    "task_id_1": true,
    "task_id_2": false,
  }
}

// ScheduleTasksModal. Existence in the set indicates the task has been selected
Set([
  "task_id_1",
  "task_id_2",
])

// ConfigurePatchCore. True indicates the task has been selected
{
  "example_bv": {
    "task1": true,
    "task2": false,
  }
}

Since the goal of this work is to include the number of generated tasks when estimating how many tasks will get scheduled in the UI, the new generatedTaskCounts resolver returns (currently a map) of tasks to their expected number of generated tasks they will create. For tasks in versions, this is relatively straightforward because we can just map the task ID to the task document's [EstimatedNumActivatedGeneratedTasks](https://github.com/evergreen-ci/evergreen/blob/8f3341c54f8430a9be6e5ae817c81a1a65f0e9b2/model/task/task.go#L302) field. Ex: 
```
{
  "task_id_1": 10,
  "task_id_2": 5,
}
```
There is slightly more complexity in the "Configure Tasks" component because we are potentially creating tasks that do not exist until we hit "Configure", so we cannot directly create a map with task IDs since there are no IDs yet. This is why the patch resolver for this field returns a map that looks like this:
```
{
  "taskName1-variantName": 10,
  "taskName2-variantName": 5
}
```
In either case, though, the data model is a map of String-> Int. It may be beneficial to use something stricter like a [StringMap](https://github.com/evergreen-ci/evergreen/pull/8169#discussion_r1714084091) but for integers instead.

Reply:
Sorry for complicating this and for the slow response. My primary concern is improving readability and ensuring we maintain type safety across the codebase. My suggestion at Introducing the `GeneratedTaskCountResults` type was aimed at reducing the need for `any` types and providing a more structured and consistent way to handle the generated task counts. My goal with introducing `GeneratedTaskCountResults` as a stricter type was to limit the ambiguity and potential errors that arise from using more generic structures like `Map` or `StringMap`.
This would help simplify the logic, making it easier to follow and reducing the cognitive load when working with these counts. It would also make the code more self-documenting, as the fields we reference would be explicitly defined rather than inferred from the map's keys. 
I do understand why this function needs to be implemented this way due to all of the different underlying implementations but I think we could still strike a balance between having a flexible implementation and a well documented type-safe one.

Reply:
Apologies for late reply as well due to outage work. I understand the concern and I've opened https://github.com/evergreen-ci/evergreen/pull/8268 on the Evergreen side and will update this PR accordingly after it is merged.